compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1715,6 +1717,33 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                         };
 
                         err.note(format!("{loop_type} evaluate to unit type `()`"));
+
+                        struct RetExprChecker {
+                            contains_ret_expr: bool,
+                        }
+                        impl<'v> Visitor<'v> for RetExprChecker {
+                            fn visit_expr(&mut self, ex: &'v Expr<'v>) {
+                                if let hir::ExprKind::Ret(_) = ex.kind {
+                                    self.contains_ret_expr = true;
+                                    return;
+                                }
+                                intravisit::walk_expr(self, ex);
+                            }
+                        }
+                        let mut visitor = RetExprChecker { contains_ret_expr: false };
+                        intravisit::walk_block(&mut visitor, loop_blk);
+                        if visitor.contains_ret_expr {
+                            let loop_header = if loop_src == hir::LoopSource::While {
+                                "loop condition expression or pattern"
+                            } else {
+                                "iterator expression"
+                            };
+                            err.span_note(
+                                    loop_span,
+                                    format!("It is determined that this might iterate zero times, regardless of the {loop_header}.")
+                                );
+                            err.span_help(loop_span, "If you are assuming that it will iterate more than once, consider using a `loop` expression instead.");
```

Comment:
```suggestion
                            err.span_help(loop_span, "If you are assuming that it will iterate at least once, consider using a `loop` expression instead.");
```